### PR TITLE
Use cache for packages independent from ocaml version

### PR DESCRIPTION
**Pull request by panglesd**
Fix "regression" introduced in #115.

We need to verify that the packages are taken first from `dependent_repo` and then from `independent_repo`.

### CODE REVIEW

**Comment by panglesd** [comment]
> We need to verify that the packages are taken first from `dependent_repo` and then from `independent_repo.`

Seems from https://opam.ocaml.org/doc/man/opam-repository.html that given the order on which the repo are added, it will be the case.

File: src/lib/tools.ml [hunk 1]
```diff
@@ -159,6 +158,9 @@ let install opam_opts tools =
     Result.List.fold_left
       (fun ((to_build, to_install, not_installed) as acc) tool ->
         let { name; pure_binary; ocaml_version_dependent; _ } = tool in
+        let repo =
+          if ocaml_version_dependent then dependent_repo else independent_repo
```
Comment (Julow) [inline src/lib/tools.ml:162]: That would be too easy. We probably shouldn't build "independent" packages on a pinned compiler to put them into the cache.
It is reasonable to pull from the cache but I'm not sure about pushing to it.

Comment (panglesd) [reply src/lib/tools.ml:162]: Very right!

File: src/lib/tools.ml [hunk 2]
```diff
@@ -132,21 +137,15 @@ let get_cache_repo opam_opts ~pinned f =
     @@ OS.Dir.with_tmp "ocaml-platform-pinned-cache-%s"
          (fun tmp_path () ->
            let* repo = Binary_repo.init tmp_path in
-           f repo)
+           f repo global_repo)
```
Comment (Julow) [inline src/lib/tools.ml:140]:
```suggestion
           f global_repo repo)
```
It's "pull repo" and "push repo"

File: src/lib/tools.ml [hunk 3]
```diff
@@ -220,13 +219,14 @@ let install opam_opts tools =
...
-        Installed_repo.with_repo_enabled opam_opts repo (fun () ->
-            let* () = Installed_repo.update opam_opts repo in
-            Logs.app (fun m -> m "* Installing tools...");
-            Opam.install
-              { opam_opts with log_height = Some 10 }
-              tools_to_install)
+        let pull_repo = Binary_repo.repo pull_repo
+        and push_repo = Binary_repo.repo push_repo in
+        Installed_repo.with_repo_enabled opam_opts pull_repo @@ fun () ->
+        Installed_repo.with_repo_enabled opam_opts push_repo @@ fun () ->
```
Comment (Julow) [inline src/lib/tools.ml:225]: What's the order in which Opam will look at the repository ? In the order they are enabled ? We want the `pull_repo` to be looked first. 

It doesn't seem to cause problems but it would be cleaner if we'd enable only the necessary repositories. The list should be defined in `get_cache_repo` too. For example:

```ocaml
let with_repos_enabled repos f =
  List.fold_left
    (fun k repo () ->
      let repo = Binary_repo.repo repo in
      Installed_repo.with_repo_enabled opam_opts repo @@ fun () ->
      let* () = Installed_repo.update opam_opts repo in
      k ())
    f repos ()
in
```

Comment (Julow) [reply src/lib/tools.ml]: I think the name are fine. "local" and "global" are too vague.

> In case of pinned compiler, we want to first look if something specific has been built for the pinned compiler (in the push_repo) and if not (for ocaml version independent packages already in the cache) look for the package in the global repo.

For pinned compilers, the "push repo" is empty, never look into it.

**Comment by panglesd** [comment]
I hope this time I got it right:
- there is `independent_repo`, a repo where to _look for_ binary package independent from the ocaml version
- and `repo`, the cache repo which might be the global cache or an empty cache for a pinned compiler. We both look for package in there and push built packages.

File: src/lib/tools.ml [hunk 4]
```diff
@@ -132,21 +137,16 @@ let get_cache_repo opam_opts ~pinned f =
     @@ OS.Dir.with_tmp "ocaml-platform-pinned-cache-%s"
          (fun tmp_path () ->
            let* repo = Binary_repo.init tmp_path in
-           f repo)
+           f global_repo repo [ global_repo; repo ])
```
Comment (Julow) [inline src/lib/tools.ml]: You swapped the order in the other function again !

I added a commit to fix that definitely.

File: test/tests/check_cache_is_sufficiently_empty.sh [hunk 5]
```diff
@@ -0,0 +1,7 @@
+#/usr/bin/env bash
+set -euo pipefail
+
+[[ $(find $HOME/.opam/plugins/ocaml-platform/cache/archives/odoc+bin+platform) = "" ]]
+[[ $(find $HOME/.opam/plugins/ocaml-platform/cache/archives/dune) = "" ]]
+[[ $(find $HOME/.opam/plugins/ocaml-platform/cache/archives/merlin+bin+platform) = "" ]]
+[[ $(find $HOME/.opam/plugins/ocaml-platform/cache/archives/ocaml-lsp-server+bin+platform) = "" ]]
```
Comment (Julow) [inline test/tests/check_cache_is_sufficiently_empty.sh:7]: Why this change ? "version independent" packages shouldn't have been cached in this test, which call the installer only once.

Comment (panglesd) [reply test/tests/check_cache_is_sufficiently_empty.sh]: You are right, no package should be put in cache. However, I think the CI is wrong as the cache repository is still created, even though no package were added. So let's check emptyness by checking that not package at all are added, instead of no cache repo created.
